samples: sensors: improve filtering

nashif opened the pull request:
Improve filtering by avoiding wildcard filters and dts lookups, instead,
target the platforms in the README or those with overlays.

Improve integration_platforms and reduce the number of platforms we
enable in integration mode.

Various other cosmetics fixes.

This will help speed up CI and avoud useless builds trying to find a
matching system when we know exactly what platforms already support
this sensor for example.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: sensors: improve filtering

Improve filtering by avoiding wildcard filters and dts lookups, instead,
target the platforms in the README or those with overlays.

Improve integration_platforms and reduce the number of platforms we
enable in integration mode.

Various other cosmetics fixes.

This will help speed up CI and avoud useless builds trying to find a
matching system when we know exactly what platforms already support
this sensor for example.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

teburd commented on samples/sensor/bme280/sample.yaml:
```diff
@@ -17,19 +17,18 @@ tests:
   sample.sensor.bme280.spi:
     harness: console
     tags: sensors
-    depends_on:
```
this should be looking for a bme280 instead, e.g. dt_compat_enable("bosch,bme280"); and then using the appropriate DT macro to find the first bme280 on a SPI bus imo, setting platform_allow to ruuvitag here looks wrong.

nashif replied:
this scenario is not buiulding for anything right now, it was wrong already, adding `ruuvi_ruuvitag` which has the sensor on SPI finally get it to build:

```
i9:bme280(a64.ld_fix): west twister -s sample.sensor.bme280.spi -v --all
Renaming output directory to /home/nashif/zephyrproject/zephyr/samples/sensor/bme280/twister-out.1
INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-2153-g31f3bf28a572
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting all possible platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/samples/sensor/bme280/twister-out/testplan.json
INFO    - JOBS: 20
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue

INFO    - 1 test scenarios (822 test instances) selected, 822 configurations skipped (822 by static filter, 0 at runtime).
INFO    - 0 of 822 test configurations passed (0.00%), 0 failed, 0 errored, 822 skipped with 0 warnings in 5.37 seconds
INFO    - In total 0 test cases were executed, 822 skipped on 822 out of total 823 platforms (99.88%)
INFO    - 0 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/samples/sensor/bme280/twister-out/twister.json
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/samples/sensor/bme280/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/samples/sensor/bme280/twister-out/twister_report.xml...
INFO    - Run completed
```

with this change....

```
820/820 ruuvi_ruuvitag            samples/sensor/bme280/sample.sensor.bme280.spi     PASSED (build)
```

nashif replied:
using filter means I have to go through all boards and see which one does have this sensor, wasting lots of resources. This is just a sample, if we know which platform has this configuration and sensor, we should not use filter, instead just allow that platform.

teburd replied:
Is this because of the whole dt requires a CMake run and build target built before seeing the full devicetree thing I ran into awhile back? Ugh

nashif replied:
yes

teburd commented on samples/sensor/grow_r502a/sample.yaml:
```diff
@@ -11,4 +11,5 @@ tests:
     build_only: true
     depends_on: serial
-    filter: dt_compat_enabled("hzgrow,r502a")
+    platform_allow:
```
Why change this from dt_compat to platform_allow? Seeing this a few times now

teburd replied:
Yeah I get it now. Avoiding DTS makes sense to me with this insight.

jeppenodgaard replied:
Never thought about that. Might this information be worth adding to https://docs.zephyrproject.org/latest/develop/test/twister.html?

teburd reviewed (commented):
small typo in commit message s/avoud/avoid

I think I get the desire behind setting explicit platform_allow to limit CI coverage to those platforms is that right?

nashif commented on samples/sensor/accel_polling/sample.yaml:
```diff
@@ -3,14 +3,17 @@ sample:
   sample.sensor.accel_polling:
     harness: console
-    tags: sensors
-    filter: dt_alias_exists("accel0")
```
you do not need the filter if you have platform_allow already that covers platforms with this alias.

MaureenHelm replied:
I really don't want to go back to platform_allow again; valid board configurations regularly got missed and broke. Would it help use the `depends on`/`supports` yaml properties instead?

jeppenodgaard commented on samples/sensor/tmp116/sample.yaml:
```diff
@@ -3,8 +3,7 @@ sample:
   sample.sensor.tmp116:
     harness: console
-    platform_allow: nucleo_f401re
-    integration_platforms:
```
Why is this removed?

nashif replied:
no need to add integration_platforms if this sample is already targeting one platform only.

PerMac commented:
I made an RFC where we can discuss how to speed up CIs without artificially limiting scopes with platform_allow https://github.com/zephyrproject-rtos/zephyr/issues/78403

nashif committed:
samples: sensors: minor cleanups to filters

Move some of the filters to lists and add integration platforms.

In cases where the sample targets one platform, no need for both
platform_allow and integration_tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: sensor: accel: reduce integration_platforms

integration_platforms is supposed to pick a small set, not list all
possible supported boards...

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
boards: twister: add supported features to boards

Extend board yaml file used by twister with additional supported
features used for testing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: sensor: bme280: fix filter

This sample was not building anywhere, target the boards directly that
provide this sensor.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
CI failure unrelated afaik

nashif commented:
@MaureenHelm  @teburd  please revisit

MaureenHelm commented on samples/sensor/accel_polling/sample.yaml:
```diff
@@ -3,26 +3,15 @@ sample:
-      - stm32f411e_disco                # lsm303agr_accel
       - stm32f3_disco                   # lsm303dlhc_accel
-      - bl5340_dvk/nrf5340/cpuapp       # lis3dh
-      - b_l4s5i_iot01a                  # lsm6dsl
-      - sensortile_box                  # lis2dw12, lsm6dso, iisdhhc
-      - thingy53/nrf5340/cpuapp         # adxl362, bmi270
+      - cc1352r_sensortag               # adxl362
```
This isn't trying to list all supported boards; it's selecting set of boards that cover different accelerometer drivers. Some duplicates slipped in along the way. This should be:
```suggestion
      - frdm_kl25z                      # mma8451q
      - lpcxpresso55s28                 # mma8652fc
      - bbc_microbit                    # mmc8653fc
      - frdm_k64f                       # fxos8700
      - mimxrt1040_evk                  # fxls8974
      - sparkfun_thing_plus/nrf9160     # lis2dh
      - thingy52/nrf52832               # lis2dh12
      - stm32f411e_disco                # lsm303agr_accel
      - stm32f3_disco                   # lsm303dlhc_accel
      - bl5340_dvk/nrf5340/cpuapp       # lis3dh
      - b_l4s5i_iot01a                  # lsm6dsl
      - sensortile_box                  # lis2dw12, lsm6dso, iisdhhc
      - thingy53/nrf5340/cpuapp         # adxl362, bmi270
```

nashif replied:
integration_platforms is supposed to list a small subset (1 or 2) for coverage and not be an exhaustive list of anything that can build, which is the case here. All supported platform will then be covered used the filter. 

see https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#twister-should-be-able-to-build-every-sample

"Make use of integration_platforms to limit the scope when there are timing or resource constraints. Ideally, one platform should be enough to verify the sample when changes are submitted to the Zephyr tree via a pull-request."

also

https://docs.zephyrproject.org/latest/develop/test/twister.html

> integration_platforms: <YML list of platforms/boards>
This option limits the scope to the listed platforms when twister is invoked with the --integration option. Use this instead of platform_allow if the goal is to limit scope due to timing or resource constraints.

We are not verifying platforms here, we are verifying the sample and that does not need to be done on each platform with this sensor.

MaureenHelm replied:
Not trying to verify platforms. We're verifying the sample builds for different sensors (that's what the comment after each platform is about)

nashif replied:
right, but integration_platforms is about limiting the scope, not adding every single possible hardware, the coverage and inclusion for the various sensors should be done in some other way

MaureenHelm commented on boards/96boards/wistrio/96b_wistrio.yaml:
```diff
@@ -12,3 +12,4 @@ supported:
   - i2c
   - eeprom
   - lora
+  - lis2dh
```
Instead of the specific sensor driver implementation, I think this should be the sensor class. e.g., `accel`.

nashif replied:
and how the pairing will work? for example `samples/sensor/lis2dh/` in this case will be built on all boards with `accel` even if it is a different sensor? That is not how most board do this right now, we point to the actual implementation.

MaureenHelm replied:
We need to remove the remaining driver-specific samples. The pairing should reflect interfaces, like for other driver classes. I2C for example, we say is supported as a driver class or interface, not as a specific implementation.

nashif replied:
but that does not reflect the state of things in the tree right now, sensors are added as the implementation and not the interface, see `boards/st/disco_l475_iot1/disco_l475_iot1.yaml`

```
supported:
  - arduino_gpio
  - arduino_i2c
  - i2c
  - hts221
  - lps22hb
  - lsm6dsl
  - pwm
  - counter
  - gpio
  - ble
  - spi
  - nvs
  - vl53l0x
  - watchdog
  - adc
  - dac
  - qspi
  - dma
  - rtc
  - usbd
  ```

We can improve this and do something like we do with network, i.e. 

`sensor:accel:lis2dh` or something along those lines, but this PR changes based on what we have right now, such an enhancement can be done as a followup.

MaureenHelm commented on boards/croxel/croxel_cx1825/croxel_cx1825_nrf52840.yaml:
```diff
@@ -14,4 +14,5 @@ supported:
   - gpio
   - i2c
   - watchdog
+  - hts221
```
Likewise, this should be:
```suggestion
  - ambient_temp
  - humidity
```

nashif replied:
ditto

MaureenHelm commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -8,7 +8,12 @@ common:
 tests:
   sample.sensor.qdec_sensor:
-    filter: dt_alias_exists("qdec0")
+    platform_allow:
```
`platform_allow` makes it impossible for twister to exercise this sample on an out-of-tree board with a qdec sensor.

nashif replied:
and why do want to do that? we are verifying samples here, this is not a test and the sample should be verified on the document and available boards in the tree with this sensor. Relying on the filter means that I have to run cmake > 800 times  just to see which board has qdec0 in their dts

something like that...

```
i9:zephyr(main): ./scripts/twister -T samples/sensor/qdec --all -v
Renaming output directory to /home/nashif/zephyrproject/zephyr/twister-out.2
INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-4713-g091c6664c51b
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting all possible platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 20
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 2592/3420 riscv32_virtual           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2593/3420 intel_adl_rvp             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2594/3420 intel_adl_crb             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2595/3420 ek_ra4m2                  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2596/3420 nucleo_wb09ke             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2597/3420 frdm_ke17z512             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2598/3420 nrf54l15dk/nrf54l15/cpuflpr samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2599/3420 nrf54l15dk/nrf54l15/cpuflpr/xip samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2600/3420 robokit1                  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2601/3420 steval_fcu001v1           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2602/3420 mg100                     samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2603/3420 olimex_stm32_e407         samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2604/3420 olimex_esp32_evb/esp32/appcpu samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2605/3420 esp32c3_devkitm           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2606/3420 stm32l152c_disco          samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2607/3420 stm32l1_disco             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2608/3420 titanium_ti60_f225        samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2609/3420 olimex_esp32_evb/esp32/procpu samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2610/3420 esp32s3_devkitm/esp32s3/appcpu samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2611/3420 arduino_due               samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2612/3420 b_g474e_dpow1             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2613/3420 arduino_portenta_h7/stm32h747xx/m4 samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2614/3420 arduino_portenta_h7/stm32h747xx/m7 samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2615/3420 em_starterkit@2.2/emsk_em7d samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2616/3420 em_starterkit/emsk_em7d   samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2617/3420 gd32e507v_start           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2618/3420 em_starterkit/emsk_em9d   samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2619/3420 em_starterkit/emsk_em11d  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2620/3420 stm32g0316_disco          samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2621/3420 innblue22/nrf9160         samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2622/3420 mini_stm32h743            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2623/3420 innblue22/nrf9160/ns      samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2624/3420 slwrb4161a                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2625/3420 stm32l476g_disco          samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2626/3420 fvp_base_revc_2xaemv8a//smp/ns samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2627/3420 fvp_base_revc_2xaemv8a    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2628/3420 bt510                     samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2629/3420 olimex_lora_stm32wl_devkit samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2630/3420 olimex_lora_stm32wl_devkit@D samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2631/3420 nucleo_g070rb             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2632/3420 cc1352r1_launchxl         samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2633/3420 max32670evkit             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2634/3420 it8xxx2_evb               samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2635/3420 nucleo_f412zg             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2636/3420 pico_pi/mcimx7d/m4        samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2637/3420 esp32_ethernet_kit/esp32/procpu samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2638/3420 esp32_ethernet_kit/esp32/appcpu samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2639/3420 xiao_ble                  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2640/3420 slwrb4180a                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2641/3420 xiao_ble/nrf52840/sense   samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2642/3420 stm32h745i_disco/stm32h745xx/m4 samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2643/3420 stm32h745i_disco/stm32h745xx/m7 samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2644/3420 max32690fthr/max32690/m4  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2645/3420 stm32f401_mini            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2646/3420 nucleo_g0b1re             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 2647/3420 96b_meerkat96/mcimx7d/m4  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
```


and just by using default twister options, i.e. what we have in CI:

```
i9:zephyr(main): ./scripts/twister -s samples/sensor/qdec/sample.sensor.qdec_sensor   -v
Renaming output directory to /home/nashif/zephyrproject/zephyr/twister-out.4
INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-4713-g091c6664c51b
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 20
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 16/44 qemu_arc/qemu_arc_hs5x    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 17/44 qemu_arc/qemu_arc_hs/xip  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 18/44 qemu_cortex_a9            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 19/44 qemu_leon3                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 20/44 qemu_malta                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 21/44 qemu_malta/qemu_malta/be  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 22/44 qemu_arc/qemu_arc_hs6x    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 23/44 qemu_xtensa/dc233c        samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 24/44 mps2/an385                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 25/44 qemu_riscv32              samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 26/44 qemu_arc/qemu_arc_hs      samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 27/44 m2gl025_miv               samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 28/44 qemu_riscv32e             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 29/44 qemu_nios2                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 30/44 qemu_riscv64/qemu_virt_riscv64/smp samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 31/44 qemu_riscv32/qemu_virt_riscv32/smp samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 32/44 mps2/an521/cpu0           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 33/44 qemu_xtensa/dc233c/mmu    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 34/44 qemu_riscv64              samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 35/44 qemu_arc/qemu_arc_em      samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 36/44 nsim/nsim_hs/mpuv6        samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 37/44 qemu_cortex_a53           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 38/44 qemu_x86                  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 39/44 qemu_cortex_r5            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 40/44 qemu_cortex_a53/qemu_cortex_a53/smp samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 41/44 qemu_cortex_m0            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 42/44 native_sim                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 43/44 qemu_x86_64               samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 44/44 nsim/nsim_sem             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)

INFO    - 1 test scenarios (44 test instances) selected, 44 configurations skipped (15 by static filter, 29 at runtime).
INFO    - 0 of 44 test configurations passed (0.00%), 0 built (not run), 0 failed, 0 errored, 44 skipped with 0 warnings in 10.07 seconds
INFO    - In total 0 test cases were executed, 44 skipped on 44 out of total 856 platforms (5.14%)
INFO    - 0 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed

```


NOTHING get built, so we have been wasting time and resourcing running cmake on this "test" 100s of times without actually verifying anything, defeats the purpose completely.

With this change:


```
i9:zephyr(topic/samples/filters_3): ./scripts/twister -s samples/sensor/qdec/sample.sensor.qdec_sensor   -v
Renaming output directory to /home/nashif/zephyrproject/zephyr/twister-out.5
INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-4718-g3066e108c138
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 20
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 45/45 mimxrt1050_evk            samples/sensor/qdec/sample.sensor.qdec_sensor      NOT RUN (build)

INFO    - 1 test scenarios (45 test instances) selected, 44 configurations skipped (44 by static filter, 0 at runtime).
INFO    - 0 of 45 test configurations passed (0.00%), 1 built (not run), 0 failed, 0 errored, 44 skipped with 0 warnings in 11.41 seconds
INFO    - In total 0 test cases were executed, 44 skipped on 45 out of total 856 platforms (5.26%)
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```


if I remove the platform_allow in this change and restore the filter:

```
i9:zephyr(topic/samples/filters_3): ./scripts/twister -s samples/sensor/qdec/sample.sensor.qdec_sensor   -v
Renaming output directory to /home/nashif/zephyrproject/zephyr/twister-out.6
INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-4718-g3066e108c138
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 20
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 16/45 nsim/nsim_hs/mpuv6        samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 17/45 qemu_riscv64/qemu_virt_riscv64/smp samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 18/45 qemu_xtensa/dc233c/mmu    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 19/45 qemu_leon3                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 20/45 qemu_malta/qemu_malta/be  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 21/45 qemu_xtensa/dc233c        samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 22/45 m2gl025_miv               samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 23/45 nsim/nsim_sem             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 24/45 mps2/an521/cpu0           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 25/45 qemu_cortex_a53           samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 26/45 qemu_cortex_a53/qemu_cortex_a53/smp samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 27/45 qemu_x86_64               samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 28/45 qemu_nios2                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 29/45 qemu_x86                  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 30/45 mps2/an385                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 31/45 qemu_riscv64              samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 32/45 qemu_riscv32/qemu_virt_riscv32/smp samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 33/45 qemu_malta                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 34/45 qemu_cortex_m0            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 35/45 qemu_arc/qemu_arc_hs5x    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 36/45 qemu_riscv32              samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 37/45 qemu_cortex_r5            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 38/45 native_sim                samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 39/45 qemu_cortex_a9            samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 40/45 qemu_arc/qemu_arc_em      samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 41/45 qemu_riscv32e             samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 42/45 qemu_arc/qemu_arc_hs6x    samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 43/45 qemu_arc/qemu_arc_hs/xip  samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 44/45 qemu_arc/qemu_arc_hs      samples/sensor/qdec/sample.sensor.qdec_sensor      SKIPPED (runtime filter)
INFO    - 45/45 mimxrt1050_evk            samples/sensor/qdec/sample.sensor.qdec_sensor      NOT RUN (build)

INFO    - 1 test scenarios (45 test instances) selected, 44 configurations skipped (15 by static filter, 29 at runtime).
INFO    - 0 of 45 test configurations passed (0.00%), 1 built (not run), 0 failed, 0 errored, 44 skipped with 0 warnings in 17.37 seconds
INFO    - In total 0 test cases were executed, 44 skipped on 45 out of total 856 platforms (5.26%)
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed

```

same thing, we now at least build on one, but still waste time on trying to figure out who has this sensor.


and this is one of the reasons why our CI is very slow, and this is just the tip of the iceberg.

We are looking for way to improve the logic of inclusion/exclusions in tests/sample via twister but the above and many other samples contribute to a large portion why we spend so much time building in CI.

MaureenHelm replied:
Why can't we use `depends on/supports` yaml properties?

nashif commented:
@MaureenHelm  ping

nashif commented:
@MaureenHelm  ping

erwango commented:
Given the questions raised above on this topic by active reviewers and maintainers, it would be worth documenting all the possible twisters filtering/selection possibilities and why/how they should be used or preferred depending on cases, otherwise this will be an endless source of debate and confusion in PR reviews (and a similar PR will be required next year).

nashif commented:
> Given the questions raised above on this topic by active reviewers and maintainers, it would be worth documenting all the possible twisters filtering/selection possibilities and why/how they should be used or preferred depending on cases, otherwise this will be an endless source of debate and confusion in PR reviews (and a similar PR will be required next year).

the questions were answered here and in the docs: https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#definition-and-criteria

erwango commented:
> > Given the questions raised above on this topic by active reviewers and maintainers, it would be worth documenting all the possible twisters filtering/selection possibilities and why/how they should be used or preferred depending on cases, otherwise this will be an endless source of debate and confusion in PR reviews (and a similar PR will be required next year).
> 
> the questions were answered here and in the docs: https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#definition-and-criteria

You gonna find me picky, but:
- answers here don't count, as they'll be soon lost in the flow for the vast majority
- Only mention on `platform_allow` explain this limit the scope of platform not why it should be preferred vs `depends on/supports` or `filter` (they could also be used to ` limit test to the platforms the sample was actually verified on`)
- It mentions `Those platforms should be listed in the sample’s README file.`. 

I'm just pointing this so we can have documented (and explained) rules that anyone (at least the ones that also care) can refer to to help in enforcing this strategy, so that you don't have to come back on this regularly. And the counter intuitive fact that `platform_allow` should be preferred to "smarter" ways of doing would make it even more needed.
